include/deal.II/lac/trilinos_tpetra_sparse_matrix.templates.h

```diff
@@ -733,10 +859,73 @@ namespace LinearAlgebra
 
 
     template <typename Number, typename MemorySpace>
-    void
+    template <typename InputVectorType>
+    std::enable_if_t<
+      !std::is_same_v<InputVectorType, Vector<Number, MemorySpace>>>
+    SparseMatrix<Number, MemorySpace>::Tvmult(InputVectorType       &dst,
+                                              const InputVectorType &src) const
+    {
+      Assert(&src != &dst, ExcSourceEqualsDestination());
+      Assert(matrix->isFillComplete(), ExcMatrixNotCompressed());
+
+      // get the size of the input vectors:
+      const size_type dst_local_size =
+        internal::end(dst) - internal::begin(dst);
+      const size_type src_local_size =
+        internal::end(src) - internal::begin(src);
+
+      // For the dst vector:
+      // create a Kokkos::View from the dst vector
+      typename VectorType::impl_scalar_type *raw_dst =
+        reinterpret_cast<typename VectorType::impl_scalar_type *>(
+          internal::begin(dst));
+      Kokkos::View<
+        typename VectorType::impl_scalar_type **,
+        Kokkos::LayoutLeft,
+        Kokkos::Device<typename MemorySpace::kokkos_space::execution_space,
+                       typename MemorySpace::kokkos_space>>
+        kokkos_view_dst(raw_dst, dst_local_size);
+
+      // get a Kokkos::DualView
+      auto mirror_view_dst = Kokkos::create_mirror_view(kokkos_view_dst);
+      typename VectorType::dual_view_type kokkos_dual_view_dst(kokkos_view_dst,
+                                                               mirror_view_dst);
+
+      // create the Tpetra::Vector
+      VectorType tpetra_dst(matrix->getDomainMap(), kokkos_dual_view_dst);
+
+      // For the src vector:
+      // create a Kokkos::View from the src vector
+      typename VectorType::impl_scalar_type *raw_src =
+        reinterpret_cast<typename VectorType::impl_scalar_type *>(
+          const_cast<double *>(internal::begin(src)));
+      Kokkos::View<
+        typename VectorType::impl_scalar_type **,
+        Kokkos::LayoutLeft,
+        Kokkos::Device<typename MemorySpace::kokkos_space::execution_space,
+                       typename MemorySpace::kokkos_space>>
+        kokkos_view_src(raw_src, src_local_size);
```

Comment:
Good point!

I thought I had to provide the memory space of the target vector. But you are right; this is not true. We need to provide the memory space of the source vector. So we need to be way more careful here...

For most vector types, the memory space is always `Host`. The only exceptions I am aware of is the `TpetraWrappers::Vector<Number, MemorySpace>` and the `distributed::Vector<Number, MemorySpace>`.

Reply:
> Do you really need a two-dimensional `View` here?
> 
> How do you know the memory space of `InputVectorType`?

The Tpetra::Vector always expects a two-dimensional View, as far as I know.

Reply:
OK. In this case, you still have to provide extents for both dimensions in the View constructor.

Reply:
> OK. In this case, you still have to provide extents for both dimensions in the View constructor.

I am not that experienced with `Kokkos::View`, what constructor should I use here?

Reply:
It should be
```
Kokkos::View<
        typename VectorType::impl_scalar_type **,
        Kokkos::LayoutLeft,
        Kokkos::Device<typename MemorySpace::kokkos_space::execution_space,
                       typename MemorySpace::kokkos_space>>
        kokkos_view_src(raw_src, src_local_size, 1);
```
so that you provide the pointer and an argument for the extent in every dimension. This should be strictly enforced in newer `Kokkos` versions. Also, there is not much of a point to provide a full `Device` template argument. In the end, we only need the `MemorySpace`. For managed Views you could provide an execution space instance as runtime argument to make sure it uses a particular stream (and fences less) or even decide not to initialize the allocation at all. None of that really matters for unmanaged Views like here, of course.